### DISCUSSION TITLE
how to run something to generate a file and source that file

### DISCUSSION THREAD

**Discussion by hardfau1t** [nushell/nushell#9786] [Q&A] [answered]
from [book](https://www.nushell.sh/book/thinking_in_nu.html#think-of-nushell-as-a-compiled-language) it is not possible to do this, but is there any other ways that can source a init.nu and if that is not present then generate and source

```nushell
if (which starship | length) >= 1 {
    const starship_init = "~/.cache/starship/init.nu"
    if not ($starship_init | path exists) {
        print -e "currently dynamically sourcing is not supported"
        starship init nu | save $starship_init
    }
   source $starship_init 
} else {
    print -e "starship is not installed"
}
```

**Comment by fdncred** [comment, marked as answer]
One thing you could try is to split your script in 2 parts. Put the `starship init nu` part in env.nu and put the `source` part in config.nu. This way you're not generating and sourcing in the same script.

**Reply by amtoine** [reply]
@fdncred is right, this is how it's supposed to be done :yum: 

a similar solution is given in the [starship documentation](https://starship.rs/#nushell)

**Reply by amtoine** [reply]
@hardfau18 
as a rule of thumb :wink: 
- paths need to be present at parse-time when running things like `use` and `source`
- modifying the `PATH`, `NU_LIB_DIRS` or creating files in already-registed paths exposes scripts and modules to the parse-time checks
- `env.nu` is the first config file to be sourced, so this is the place of choice to expose stuff to the parser
